presentation/src/main/java/me/androidbox/presentation/login/viewmodel/LoginViewModel.kt

```diff
@@ -30,4 +46,24 @@ class LoginViewModel @Inject constructor() : ViewModel() {
     fun onPasswordVisibilityChanged() {
         isPasswordVisible = !isPasswordVisible
     }
+
+    fun loginUser(email: String, password: String) {
+        viewModelScope.launch {
+            try {
+                loginMutableState.value = NetworkResponseState.Loading
+
+                val loginModel = loginUseCase.execute(email, password)
+                loginMutableState.value = NetworkResponseState.Success(loginModel)
+            }
+            catch (httpException: HttpException) {
+                loginMutableState.value = NetworkResponseState.Failure(httpException)
+            }
+            catch (exception: Exception) {
+                if(exception is CancellationException) {
+                    throw Exception(exception.message)
+                }
+                loginMutableState.value = NetworkResponseState.Failure(exception)
+            }
+        }
+    }
```

Comment:
1) launch the coroutine in viewModelScope
2) Sets the Loading state
3) Waits for the response back
4) sets the loginModel if success
5) set the exception if failure

I am wondering where should I collect the `loginState` should it be in the `LoginScreen` or should it be in the `HomeActivity`?

Reply:
This would be the same for the `RegisterViewModel` and `RegisterScreen`

Reply:
Since you are passing the viewmodel to the LoginScreen, then you can collectAsState on the LoginScreen

Reply:
Also, exception handling for data related exceptions should happen in your data layer (in the repo), not in the ViewModel. Here, you're effectively binding implementation details to your presentation layer, since the `HttpException` comes from Retrofit. Therefore, presentation knows what you use to retrieve data, which it shouldn't in clean architecture. 
The goal is that if an implementation detail changes (e.g. your lib to make network calls), you ONLY want to change things in the data layer. Here, you'd need to change presentation as well.

Also mapping everything to `NetworkResponseState` should happen in the repository, so all the VM needs to do is to check the result from the repo and update its state based on that.